classes/migration/upgrade/v3_4_0/I7014_DoiMigration.inc.php

```diff
@@ -0,0 +1,486 @@
+{
+    /**
+     * @inheritDoc
+     */
+    public function up(): void
+    {
+        // DOIs
+        Schema::create('dois', function (Blueprint $table) {
+            $table->bigInteger('doi_id')->autoIncrement();
+            $table->bigInteger('context_id');
+            $table->string('doi');
+            $table->smallInteger('status')->default(1);
+
+            $table->foreign('context_id')->references('press_id')->on('presses');
+        });
+
+        // Settings
+        Schema::create('doi_settings', function (Blueprint $table) {
+            $table->bigInteger('doi_id');
+            $table->string('locale', 14)->default('');
+            $table->string('setting_name', 255);
+            $table->text('setting_value')->nullable();
+
+            // TODO: #doi Check on index/unique alongside foreign. Is it okay?
+            $table->index(['doi_id'], 'doi_settings_doi_id');
+            $table->unique(['doi_id', 'locale', 'setting_name'], 'doi_settings_pkey');
+            $table->foreign('doi_id')->references('doi_id')->on('dois');
+        });
+
+        // Add doiId to publication
+        Schema::table('publications', function (Blueprint $table) {
+            $table->bigInteger('doi_id')->nullable();
+            $table->foreign('doi_id')->references('doi_id')->on('dois')->nullOnDelete();
+        });
+
+        // Add doiId to chapters
+        Schema::table('submission_chapters', function (Blueprint $table) {
+            $table->bigInteger('doi_id')->nullable();
+            $table->foreign('doi_id')->references('doi_id')->on('dois')->nullOnDelete();
+        });
+
+        // Add doiId to publication formats
+        Schema::table('publication_formats', function (Blueprint $table) {
+            $table->bigInteger('doi_id')->nullable();
+            $table->foreign('doi_id')->references('doi_id')->on('dois')->nullOnDelete();
+        });
+
+        // Add doiId to submission files
+        Schema::table('submission_files', function (Blueprint $table) {
+            $table->bigInteger('doi_id')->nullable();
+            $table->foreign('doi_id')->references('doi_id')->on('dois')->nullOnDelete();
+        });
+
+        $this->migrateExistingDataUp();
+    }
+
+    /**
+     * Reverse the downgrades
+     *
+     * @throws DowngradeNotSupportedException
+     */
+    public function down(): void
+    {
+        throw new DowngradeNotSupportedException();
+    }
+
+    private function migrateExistingDataUp()
+    {
+        // Find all existing DOIs, move to new DOI objects and add foreign key for pub object
+        $this->_migrateDoiSettingsToContext();
+        $this->_migratePublicationDoisUp();
+        $this->_migrateRepresentationDoisUp();
+        $this->_migrateChapterDoisUp();
+        $this->_migrateSubmissionFileDoisUp();
+    }
+
+    /**
+     * Move DOI settings from plugin_settings to Context (Press) settings
+     */
+    private function _migrateDoiSettingsToContext()
```

Comment:
Might be able to move some of this into pkp-lib where the settings are shared across apps. Not a merge blocker but good to clean up for future us.

Reply:
This could be a little tricky because of the different context ID names. I've filed an issue (pkp/pkp-lib#7687) to take a look at this post merge.